File: algorithms/2_medium/M/03742_maximum_path_score_in_a_grid/dp_static_array.cpp
const size_t MAX_ROWS{ 200 };
const size_t MAX_COLS{ 200 };
const size_t MAX_K{ 1'000 };

class Solution {
public:
    static int maxPathScore(const vector<vector<int>>& grid, const int k) {
        const size_t rows{ grid.size() };
        const size_t cols{ grid[0].size() };
        const size_t k_usize{ static_cast<size_t>(k) };
        static std::mutex mtx;
        std::lock_guard<std::mutex> lock(mtx);
        static array<array<array<int, MAX_K + 1>, MAX_COLS>, MAX_ROWS> dp;

        for (size_t i{ 0 }; i < rows; ++i) {
            for (size_t j{ 0 }; j < cols; ++j) {
                for (size_t c{ 0 }; c <= k_usize; ++c) {
                    dp[i][j][c] = -1;
                }
            }
        }

        dp[0][0][0] = 0;

        for (size_t i{ 0 }; i < rows; ++i) {
            for (size_t j{ 0 }; j < cols; ++j) {
                const int score_cell{ grid[i][j] };
                const int cost_cell{ score_cell != 0 ? 1 : 0 };

                if (i > 0) {
                    for (size_t c{ 0 }; c <= k_usize; ++c) {
                        if (dp[i - 1][j][c] != -1) {
                            if (c + cost_cell <= k_usize) {
                                dp[i][j][c + cost_cell] = std::max(dp[i][j][c + cost_cell], dp[i - 1][j][c] + score_cell);
                            }
                        }
                    }
                }

                if (j > 0) {
                    for (size_t c{ 0 }; c <= k_usize; ++c) {
                        if (dp[i][j - 1][c] != -1) {
                            if (c + cost_cell <= k_usize) {
                                dp[i][j][c + cost_cell] = std::max(dp[i][j][c + cost_cell], dp[i][j - 1][c] + score_cell);
                            }
                        }
                    }
                }
            }
        }

        return *std::max_element(dp[rows - 1][cols - 1].cbegin(), dp[rows - 1][cols - 1].cbegin() + k_usize + 1);
    }
};

File: algorithms/2_medium/M/03742_maximum_path_score_in_a_grid/dp_static_array.rs
use std::sync::Mutex;

impl Solution {
    const MAX_ROWS: usize = 200;
    const MAX_COLS: usize = 200;
    const MAX_K: usize = 1_000;

    pub fn max_path_score(grid: Vec<Vec<i32>>, k: i32) -> i32 {
        let rows = grid.len();
        let cols = grid[0].len();
        let k_usize = k as usize;
        static DP: Mutex<[[[i32; Solution::MAX_K + 1]; Solution::MAX_COLS]; Solution::MAX_ROWS]> = {
            Mutex::new(
                [[[-1_i32; Solution::MAX_K + 1]; Solution::MAX_COLS]; Solution::MAX_ROWS]
            )
        };
        let mut dp = DP.lock().unwrap();
        for i in 0..rows {
            for j in 0..cols { 
                for c in 0..=k_usize {
                    dp[i][j][c] = -1;
                }
            }
        }
        dp[0][0][0] = 0;

        for i in 0..rows {
            for j in 0..cols {
                let score_cell = grid[i][j];
                let cost_cell = if score_cell != 0 { 1_usize } else { 0_usize };

                if i > 0 {
                    for c in 0..=k_usize {
                        if dp[i - 1][j][c] != -1 {
                            if c + cost_cell <= k_usize {
                                dp[i][j][c + cost_cell] = dp[i][j][c + cost_cell].max(dp[i - 1][j][c] + score_cell);
                            }
                        }
                    }
                }

                if j > 0 {
                    for c in 0..=k_usize {
                        if dp[i][j - 1][c] != -1 {
                            if c + cost_cell <= k_usize {
                                dp[i][j][c + cost_cell] = dp[i][j][c + cost_cell].max(dp[i][j - 1][c] + score_cell);
                            }
                        }
                    }
                } 
            }
        }

        dp[rows - 1][cols - 1].iter().take(k_usize + 1).map(|x| *x).max().unwrap()
    }
}
